### PR TITLE
Eunice - Batch 9 Section 8

### PR DESCRIPTION
**Pull request by Eunicesayshi**
Please fill out the survey before submitting the pull request. Thanks!

🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

How many hours did you spend on this assignment?
6 hours

Please fill in one error and/or error message you received while working on this assignment.
Reset git command (because I forgot to describe my commit) and found that my entire codes on js went missing (i.e. reverted back to original version). Thankfully I managed to get it back!

What part of the assignment did you spend the most time on?
Figuring out how the syntax work in general, such as what to write within the definition of a function to make it work like what I want to, how to call for the function later on and whether I need to assign a variable or whether I can simply use the function directly etc.

Comfort Level (1-5):
1 being very not comfortable, and 5 being comfortable: I rate it a 3.5 to 4.

Completeness Level (1-5):
5 (complete)

What did you think of this deliverable?
cool - game concept is fun.

Is there anything in this code that you feel pleased about?
My effort hahaha, presentation, scripting of the outcomes, writing out steps clearly... though I would like to also check out other methods that may be simpler! Mine may be too wordy?

### CODE REVIEW

**Comment by Eunicesayshi** [comment]
Please fill out the survey before submitting the pull request. Thanks!

🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

How many hours did you spend on this assignment?
Second attempt from scratch to familiarise myself and to attempt use of Boolean and Not in the assignment - Around 3 to 4 hours?

Please fill in one error and/or error message you received while working on this assignment.
ReferenceError: Invalid left-hand side in assignment

What part of the assignment did you spend the most time on?
-

Comfort Level (1-5):
Around 4 (rather comfortable)

Completeness Level (1-5):
5 (complete)

What did you think of this deliverable?
Fun

Is there anything in this code that you feel pleased about?
Explored putting a different logic (from my first attempt) into codes

**Comment by Eunicesayshi** [comment]
Please fill out the survey before submitting the pull request. Thanks!

🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

How many hours did you spend on this assignment?
3 hours

Please fill in one error and/or error message you received while working on this assignment.
Missing curly bracket

What part of the assignment did you spend the most time on?
This and that.. e.g. relooking at namings/labelling of variables etc.. reversing based on current logic.. figuring out how to change modes (that was before I watched the Day 5 pre-class video)

Comfort Level (1-5):
3.5 to 4.

Completeness Level (1-5):
Only the base so far, but a 5 for the base!

What did you think of this deliverable?
fun though can be tiring

Is there anything in this code that you feel pleased about?
Effort down to the details of making the word, "time", plural where needed

**Comment by Eunicesayshi** [comment]
added Korean version

File: script.js [hunk 1]
```diff
@@ -1,4 +1,586 @@
...
+
+// Create a function to change game mode based on input
+var changeGameMode = function (userInput) {
+  var gameMode = currentGameMode;
+  if (userInput == `standard`) {
+    gameMode = `standard game`;
+  }
+  if (userInput == `korean`) {
+    gameMode = `korean game`;
+  }
```
Comment (jerome-chua) [inline script.js:449]: There is something know as [SCREAM_CASE constants ](https://www.freecodecamp.org/news/when-to-capitalize-your-javascript-constants-4fabc0a4a4c4/)

Do try if you can to incorporate to project 2; it's simply another way to name variables that remain unchanged through out the program eg.

Declare these as global variables:
```
var STANDARD_MODE = "standard"
var KOREAN_MODE = "korean game"
```
Usage example:
```
  if (userInput == `korean`) {
    gameMode = KOREAN_MODE;
  }
```

File: script.js [hunk 2]
```diff
@@ -1,4 +1,586 @@
...
+//       myOutputValue = reversedGame(input, programChoice);
+//     }
+//   }
+//   return myOutputValue;
+// };
+
+//======SPS Game with Korean Game Mode: Project Part 2 (More Comfortable)=======
+// Specify global variables
+var currentGameMode = `waiting for player name`;
+var lastWonPlayer = `unassigned`;
```
Comment (jerome-chua) [inline script.js:439]: Very clear naming of variables 👍

File: script.js [hunk 3]
```diff
@@ -1,4 +1,586 @@
...
+  var randomDecimal = Math.random() * 3;
+  var randomInteger = Math.floor(randomDecimal);
+  if (randomInteger == 0) {
+    assignedChoice = `scissors`;
+  }
+  if (randomInteger == 1) {
+    assignedChoice = `paper`;
+  }
+  if (randomInteger == 2) {
+    assignedChoice = `stone`;
```
Comment (jerome-chua) [inline script.js:465]: Like wise for these,
```
var STONE = "stone"
var SCISSORS = "scissors"
var PAPER = "paper"
```

File: script.js [hunk 4]
```diff
@@ -1,4 +1,586 @@
...
+  if (userInput == programSelection) {
+    gameResult = `draw`;
+  }
+  if (
+    (userInput == `scissors` && programSelection == `paper`) ||
+    (userInput == `paper` && programSelection == `stone`) ||
+    (userInput == `stone` && programSelection == `scissors`)
+  ) {
+    gameResult = `win`;
+  }
```
Comment (jerome-chua) [inline script.js:482]: Great decomposition & clear comments, excellent! Please keep it up

File: script.js [hunk 5]
```diff
@@ -1,4 +1,586 @@
...
+    (userInput == `scissors` && programSelection == `stone`) ||
+    (userInput == `stone` && programSelection == `paper`) ||
+    (userInput == `paper` && programSelection == `scissors`)
+  ) {
+    gameResult = `lose`;
+  }
+  console.log(`user input ${userInput}`);
+  console.log(`program selection ${programSelection}`);
+  console.log(`game result ` + gameResult);
+  return gameResult;
```
Comment (jerome-chua) [inline script.js:493]: Good console.log statements for self-debugging, nice!

File: script.js [hunk 6]
```diff
@@ -1,4 +1,586 @@
...
+    (userInput == `stone` && programSelection == `scissors`)
+  ) {
+    gameResult = `win`;
+  }
+  if (
+    (userInput == `scissors` && programSelection == `stone`) ||
+    (userInput == `stone` && programSelection == `paper`) ||
+    (userInput == `paper` && programSelection == `scissors`)
+  ) {
+    gameResult = `lose`;
```
Comment (jerome-chua) [inline script.js:488]: Great that you have a clear understanding on how the if else loops runs, and this being the default result being assigned to `gameResult`

File: script.js (hunk 1 above)
Comment (jerome-chua) [reply script.js:449]: Good decomposition though!

File: script.js [hunk 7]
```diff
@@ -1,4 +1,586 @@
...
+  if (gameResult == `draw`) {
+    message = `It's a draw! <br> <br> This is the standard game. <br> <br> We both chose ${programSelection}! <br> You didn't copy my answer did you? 😜`;
+  }
+  if (gameResult == `win`) {
+    message = `You won! <br> <br> This is the standard game. <br> <br> You chose ${userInput} and I chose ${programSelection}. <br> Good job!`;
+  }
+  if (gameResult == `lose`) {
+    message = `You lost! <br> <br> This is the standard game. <br> <br> You chose ${userInput} and I chose ${programSelection}. <br> Better luck next time!`;
+  }
+  return message;
```
Comment (jerome-chua) [inline script.js:513]: This is very clear that you understand how the code is flowing, excellent work!

File: script.js [hunk 8]
```diff
@@ -1,4 +1,586 @@
...
+};
+
+// Create a function to update winning player for the last round
+var updateWinner = function (userInput, programSelection) {
```
Comment (jerome-chua) [inline script.js:517]: It is great that you are keeping the naming of your parameters - `useInput, programSelection` for your relevant functions consistent. 👍

File: script.js [hunk 9]
```diff
@@ -1,4 +1,586 @@
...
+  gameResult = result(userInput, programSelection);
+  if (gameResult == `win`) {
+    winner = `YOU are`;
+  }
+  if (gameResult == `lose`) {
+    winner = `I am`;
+  }
+  console.log(`${winner} the winner from last round.`);
+  return winner;
+};
```
Comment (jerome-chua) [inline script.js:528]: Nice decomposition once again to contain updating the winner in one function

File: script.js [hunk 10]
```diff
@@ -1,4 +1,586 @@
...
+var koreanGame = function (
+  userInput,
+  programSelection,
+  gameResult,
+  lastWonPlayer
+) {
+  var message = `Sorry, I didn't get that. <br> <br> Please type 'scissors', 'paper' or 'stone' to start playing the game with me 😊 <br> <br> You can also type 'korean' to try our korean mode and 'standard' for the original game. <br> <br> --- What is the korean version of SPS? --- <br> In Korea, a two-player upgraded version exists by the name muk-jji-ppa. After showing their hands, the player with the winning throw shouts "muk-jji-ppa!" upon which both players throw again. If they throw differently (for example, rock and paper, or paper and scissors), whoever wins this second round shouts "muk-jji-ppa!" and thus the play continues until both players throw the same item (for example, rock and rock), at which point whoever was the last winner becomes the actual winner.`;
+
+  if (userInput == `korean`) {
+    message = `You are now in the korean game mode. <br> <br> Please type 'scissors', 'paper' or 'stone' to start the korean SPS game with me! <br> <br> --- How to win? --- <br> Once both players threw the same item, whoever won the last round becomes the actual winner.`;
```
Comment (jerome-chua) [inline script.js:540]: Good thinking for the user, in breaking up the messages into two parts,
one in the original message, and the second one here

File: script.js [hunk 11]
```diff
@@ -1,4 +1,586 @@
...
+    assignedChoice = `paper`;
+  }
+  if (randomInteger == 2) {
+    assignedChoice = `stone`;
+  }
+  return assignedChoice;
+};
+
+// Create a function to derive the result of a standard SPS game
+var result = function (userInput, programSelection) {
```
[the remark this answers is not in the document]
Comment (jerome-chua) [reply script.js:471]: Nice work creating a function to output the gaming message which you then use elsewhere to print out the overall message to user! Well done!

File: script.js [hunk 12]
```diff
@@ -1,4 +1,586 @@
...
+    if (currentGameMode == `standard game`) {
+      myOutputValue = standardGame(input, programChoice);
+    }
+    if (currentGameMode == `korean game`) {
+      myOutputValue = koreanGame(
+        input,
+        programChoice,
+        gameOutcome,
+        lastWonPlayer
+      );
```
Comment (jerome-chua) [inline script.js:582]: Think these have to be indented to the right

Comment (jerome-chua) [reply script.js:582]: Keep up the way you are naming your variables. It is clear, you could explore shortening some, for eg. `currentGameMode` to `currGameMode` - this is minor though

Do explore and utilise scream case variables for upcoming projects.

Overall, I can you do have a good grasp of how the program is flowing from top to bottom, so keep it up Eunice! You're really acing 💪
